Add Configuration Files support docs (#918)
grouparoo/www.grouparoo.com: pages/docs/support/config-files.mdx

---
title: "Configuration Files"
date: "2022-02-09"
pullQuote: "Definitions and examples of Grouparoo configuration files."
---

> Grouparoo projects contain configuration files for Apps, Models, Sources, Properties, Groups, Destinations, and any other objects is exported into config files.

Reviewer: ```suggestion
Grouparoo projects contain configuration files for Apps, Models, Sources, Properties, Groups, Destinations, and any other objects generated in the course of configuration. 
```

These config files can be created and updated with [UI config](/docs/config).

> Config files are either JSON or JavaScript which allow you the flexibility to manipulate data or load configuration from external sources.

Author: I left references to the `JavaScript` files from code config. But should this be removed? In a world without generators should we only encourage JSON files?

## Config Files Location

Config files are saved at the root of your Grouparoo project directory, in a directory called `config`. A typical Grouparoo project structure looks like this:

```
my-grouparoo-project/
├── package.json
├── package-lock.json
├── .env
└── config/
    └── ... (config files go here)
```

You can optionally choose to use another directory for your configuration by setting the `GROUPAROO_CONFIG_DIR` environment variable, for example `GROUPAROO_CONFIG_DIR=/path/to/config`.

## Common Config Attributes

Every type of object has its own definition but all of them require these attributes:

- `id`: (**required**) A unique identifier for this object.
  - Lowercase letters, numbers, and underscores only
  - No spaces
  - Fewer than 38 characters
  - Must be unique among other objects of the same type
- `class`: (**required**) The type of object that is being defined such as `App`, `Model`, `Source`, `Property`, `Group`, `Destination`
- `name`: (**required**) The display name of this object

## App Config

The App configures the connections available to create Sources and Destinations.

- `id`: (**required**) Unique identifier for this app config.

> - `name`: (**required**) A display name. It matches the `id` by default.

Reviewer: Am I right to think that if the name is assigned in config UI and does not meet the requirements for ID, the name will be altered to assign a valid id? If so, do you think we should we add something like this so users aren't confused about rules for assigning a name in the UI if they read this?

Author: I'll remove this because it's related to the generator. In config UI, the name is driven by what the user types and this is more related to how the generator worked

- `class`: (**required**) This must be set to `"App"`.

> - `type`: (**required**) The type of the app define by a plugin. While the type can match the name of the plugin, plugins may offer multiple apps and the names may not match.

Reviewer: ```suggestion
- `type`: (**required**) The type of the app defined by a plugin. While the type can match the name of the plugin, plugins may offer multiple apps and the names may not match.
```

- `refresh`: App refresh query. See below for details.
- `options`: The App configuration options. These vary depending on the type of app.

**Example:**

```JSON
{
  "class": "App",
  "id": "data_warehouse",
  "name": "data_warehouse",
  "type": "postgres",
  // App Refresh Query
  "refresh": {
    "query": "..."
    "recurringFrequency": 6000
  }
  // Option vary depending on the type of app
  // Below is an example of the options for the Postgres app
  "options": {
    "host": "localhost",
    "port": 5432,
    "database": "...",
    "schema": "public",
    "user": "...",
    "password": "...",
    "ssl": false,
  }
}
```

**App Refresh Query Configuration**

App Refresh Query enables Grouparoo to check your data warehouse for new data, and run all Schedules for that App if it finds any. The following Plugins support App Refresh Query:

- BigQuery
- ClickHouse
- MySQL
- Postgres
- Redshift
- Snowflake
- SQLite

> For the best results, configure your query up to return a single row of data. For example, you might set it up as:

Reviewer: ```suggestion
For the best results, configure your query to return a single row of data. For example, you might set it up as:
```

```JSON
  "refresh": {
    "query": "SELECT MAX(updated_at) FROM meta_table"
    "recurringFrequency": 6000
  }
```

This query would check every minute for a new `updated_at` value in a table called `meta_table`. The table you check with your App Refresh Query may or may not be one that you include as a [Source](#source-config).

If there are [Schedules](#schedule-config) on this App's Sources that you do not want to be run, you can add `"refreshEnabled": false` to your Schedule's config.

## Model Config

The Model config file is currently very basic:

- `class`: (**required**) Should be set to `"Model"`.
- `id`: (**required**) The unique value that identifies this Model.
- `name`: (**required**) A display name for the Model.
- `type`: (**required**) Can be set to `"profile"`, `"account"`, or `"custom"`.

**Example:**

```JSON
{
  "class": "Model",
  "id": "users",
  "name": "Users",
  "type": "profile"
}
```

## Source Config

> Each Source has a common set of options, while other options are specific to the _type_ of Source. Here are the common options:

Reviewer: ```suggestion
Each Source has a common set of options. Certain _types_ of Sources may have additional options, but all Sources will have the following common options:
```

- `"exists"` - exists with any value (is not null)
- `"notExists"` - does not exist with any value (is null)
- `"eq"`- equals
- `"ne"` - does not equal
- `"gt"` - greater than
- `"gte"` - greater than or equal to
- `"lt"` - less than
- `"lte"` - less than or equal to
- `"substring"` - contains
- `"notSubstring"` - does not contain
- `"in"` - in a given set

For example, to create a schedule that will run every 5 minutes checking the `"updated_at"` column for rows where `status = "active"` and `last_purchase"` exists (is not null), you could do the following:

```json
{
  "class": "Schedule",
  "id": "users_schedule",
  "name": "Users Schedule",
  "sourceId": "users",
  "recurring": true,
  "recurringFrequency": 5 * 60 * 1000,
  "confirmRecords": false,
  "options": {
    "column": "updated_at"
  },
  "filters": [
    {
      "key": "status",
      "op": "eq",
      "match": "active"
    },
    {
      "key": "last_purchase",
      "op": "exists"
    }
  ]
}
```

If you've configured an App Refresh Query when you configured your App, you can also opt to leave a particular Schedule out of its updates by adding `"refreshEnabled": false` to the Schedule's config.

## Property Config

> Each Property has a common set of options, while other options are specific to the _type_ of Property. Here are the common attributes:

Reviewer: ```suggestion
Each Property has a common set of options. Certain _types_ of Properties may have additional options, but all Properties will have the following common options:
```

## Group Config

Group config files have a common format.

- `id`: (**required**) The unique value that identifies this Source.
- `class`: (**required**) Should be set to `"Group"`.
- `modelId`: (**required**) The model that the group belongs to.
- `name`: (**required**) A display name for the Group.
- `rules`: (**required**) An array of rules to filter Records. See more details below.

**Example:**

```JSON
{
  "id": "high_value_accounts",
  "class": "Group",
  "modelId": "accounts",
  "name": "High Value Accounts",
  "rules": [
    {
      "propertyId": "account_value",
      "op": "gt",
      "match": "800"
    }
  ]
}
```

Configuring a Group is all about adding Rules to the Group. A Rule is some logical operator to Filter Properties. A Record must match every Rule to be included in the Group.

Available operations depends on both the _type_ of Property and the underlying Grouparoo database (SQLite vs. Postgres). See below for a list of rules. (If in doubt, choose Postgres.)

Note you can use SQL wildcard characters (%, \_, [ ], ^, -) in your Group rule match values as appropriate and necessary.

<RuleOpsTable />

## Destination Config

> Each Destination has a common set of options, while other options are specific to the _type_ of Destination. Here are the commons set of options that may look similar to other config object shapes in Grouparoo:

Reviewer: ```suggestion
Each Destination has a common set of options. Certain _types_ of Destinations may have additional options, but all Destinations will have the common options. These common options, which may look very similar to other config object shapes in Grouparoo, are listed here: 
```

- `mapping`: See details below.
- `destinationGroupMemberships`: See details below.

**Example:**

```JSON
[
  {
    "id": "newsletter",
    "name": "newsletter",
    "class": "destination",
    "type": "mailchimp-export",
    "appId": "...",
    "collection": "...",
    "groupId": "...",
    "syncMode": "...",
    "options": {
      "listId": "...",
    },
    "mapping": {
      "email_address": "email",
      "FNAME": "firstName",
      "LNAME": "lastName",
    },
    "destinationGroupMemberships": {
      "High Value Customers!": "highValueCustomers",
    },
  }
]
```

**Destination Mappings**

Mappings are a series of key-value pairs that tell Grouparoo which Properties to export to the Destination for each Record. The structure may vary for each type of Destination, but generally, the keys are the fields in the Destination, while the values are the `id` values for the Properties in Grouparoo.

Here's an example of a Mailchimp mapping configuration:

```json
{
  "mapping": {
    "email": "email",
    "FNAME": "firstName",
    "LNAME": "lastName"
  }
}
```

Notice the odd structure of the keys. That's specific to Mailchimp.

**Destination Group Membership**

> Grouparoo also provides the ability to attach Group membership to the exported data. How this looks within the Destination is specific to the Destination. But the shape of the config file largely remains the same. In most Destinations, Destinations Group Memberships become either tags (Mailchimp), lists (Hubspot), or appended as additional properties to the Record (Customer.io).

Reviewer: ```suggestion
Grouparoo also provides the ability to attach Group membership to the exported data. How this looks within the Destination is specific to the Destination. But the shape of the config file largely remains the same. In most Destinations, Destination Group Memberships either become tags (Mailchimp) or lists (Hubspot), or are appended as additional properties to the Record (Customer.io).
```

```json
{
  "destinationGroupMemberships": {
    "High Value Customers!": "highValueCustomers"
  }
}
```

In the example above, the key represents the tag in Mailchimp, while the value is the `id` of the Group in Grouparoo.

## Using Dynamic Values

There are two methods for loading configuration information dynamically:

- Environment Variables
- JavaScript Functions

**Environment Variables**

If you want to set a variable called `production_hubspot_api_key` for use only in Hubspot Apps, you would set `GROUPAROO_OPTION__APP__production_hubspot_api_key=abc123`.

<Alert variant="warning">
  <p className="mb-0">
    ⚠️ <code>APP</code> (the object type) is surrounded by double underscores.
  </p>
</Alert>

When setting an Environment Variable to use with Grouparoo, you'll be using the pattern:

```bash
export GROUPAROO_OPTION__{topic}__{name}={value}
```

See [the Secrets page](/docs/deployment/secrets) for more information about using environment variables for config.

**JavaScript Functions**

> Code config files can be use JavaScript functions that can export an `async` method, which can `await` while loading data from an database or API. (See the _Examples_ section below.)

Reviewer: ```suggestion
Code config files can use JavaScript functions that can export an `async` method, which can `await` while loading data from a database or API. (See the _Examples_ section below.)
```
